File: python/problems/smallest-subarray.py
from bisect import bisect_left,bisect_right

def Merge(left,right):
    mergedList = []
    i = j = 0
    
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            mergedList.append(left[i])
            i+=1
        else:
            mergedList.append(right[j])
            j+=1
    while i < len(left):
        mergedList.append(left[i])
        i+=1
    while j < len(right):
        mergedList.append(right[j])
        j+=1
    
    return mergedList
def BuildMergeSortTree(Tree,A,index,start,end):
    if start == end:
        Tree[index].append(A[start])
        return
    else:
        mid = (start+end) // 2
        BuildMergeSortTree(Tree,A,2*index,start,mid)
        BuildMergeSortTree(Tree,A,2*index+1,mid+1,end)
        Tree[index] = Merge(Tree[2*index],Tree[2*index+1])
def query(Tree,index,start,end,left,right,k):
    if right < start or left > end :
        return 0
    if start >= left and end <= right:
        # binary search to find index of k
        print(Tree[index])
        try:
            start_ele = Tree[index][0]
        except:
            start_ele = 0
        return len(Tree[index]) - (bisect_left(Tree[index], k)) - start_ele
    mid = (start+end) //2
    return query(Tree,2*index,start,mid,left,right,k)+query(Tree,2*index+1,mid+1,end,left,right,k)
     
def solve(A,B,N):
    Tree = [[] for i in range(4*N+1)]
    BuildMergeSortTree(Tree,A,0,0,N-1)
    for i in range(N):
        value = A[i]
        start = i
        end = N-1
        ans = -1
        while start <= end:
            m = (start + end) >> 1
            cnt = query(Tree,1,0,N-1,i,m,value)
            if cnt >= B[i]:
                ans = m
                end = m-1
            else:
                start = m+1
        if ans == -1:
            print(-1,end = ' ')
        else:
            print(ans-i+1,end = ' ') 
        
N = 5
A = list(map(int,'8 2 4 1 9'.split()))
B = list(map(int,'2 3 5 1 1'.split()))     
solve(A,B,N)

File: cpp/range-queries/smallest-subarray.cpp
#include<bits/stdc++.h>
#define ll long long int
using namespace std;

vector<int> merge(vector<int>& v1, vector<int>& v2) 
{ 
    int i = 0, j = 0; 
   
    vector<int> v; 
  
    while (i < v1.size() && j < v2.size()) { 
        if (v1[i] <= v2[j]) { 
            v.push_back(v1[i]); 
            i++; 
        } 
        else { 
            v.push_back(v2[j]); 
            j++; 
        } 
    } 
  
    for (int k = i; k < v1.size(); k++) 
        v.push_back(v1[k]); 
    for (int k = j; k < v2.size(); k++) 
        v.push_back(v2[k]); 
    return v; 
} 
  
void buildTree(vector<int>* tree, int* arr, 
               int index, int s, int e) 
{ 
  
    if (s == e) { 
        tree[index].push_back(arr[s]); 
        return; 
    } 
  
    int mid = (s + e) / 2; 
    buildTree(tree, arr, 2 * index, s, mid); 
    buildTree(tree, arr, 2 * index + 1, mid + 1, e); 
  
    tree[index] = merge(tree[2 * index], tree[2 * index + 1]); 
} 
  
int query(vector<int>* tree, int index, int s, 
          int e, int l, int r, int k) 
{ 
  
    if (r < s || l > e) 
        return 0; 
  
    if (s >= l && e <= r) { 
        return (tree[index].size() 
                - (lower_bound(tree[index].begin(), 
                               tree[index].end(), k) 
                   - tree[index].begin())); 
    } 
  
    int mid = (s + e) / 2; 
    return (query(tree, 2 * index, s, 
                  mid, l, r, k) 
            + query(tree, 2 * index + 1, mid + 1, 
                    e, l, r, k)); 
} 
  

void solve()
{
  int n;
  cin >> n;
  assert(1 <= n and n <= 300000);

  int a[n];
  for(int i = 0 ; i < n ; i++)
  { 
    cin >> a[i];
    assert(1 <= a[i] and a[i] <= 1000000000);
  }

  int b[n];
  for(int i = 0 ; i < n ; i++)
  {
    cin >> b[i];
    assert(1 <= b[i] and b[i] <= n);
  }

  vector<int> tree[4*n+1];
  buildTree(tree, a, 1 , 0 , n - 1);

  for(int i = 0 ; i < n ; i++)
  {
    int value = a[i];
    int s = i;
    int e = n - 1;
    int ans = -1;

    while(s <= e){
      int m = (s+e) >> 1;
      int cnt = query(tree, 1, 0, n - 1, i, m, value);

      if(cnt >= b[i])
      {
        ans = m;
        e = m - 1;
      }
      else{
        s = m + 1;
      }
    }

    if(ans == -1) cout << ans << " ";
    else cout << (ans - i + 1 ) << " ";
  }
  cout << endl;
} 

int main(){
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  int t;
  t = 1;
  while(t--)
  {
    solve();
  }
}